packages/forklift-console-plugin/src/modules/Providers/views/details/tabs/VirtualMachines/utils/helpers/getVmPowerState.ts

```diff
@@ -0,0 +1,58 @@
+import {
+  OpenshiftVM,
+  OpenstackVM,
+  OVirtVM,
+  ProviderType,
+  ProviderVirtualMachine,
+  VSphereVM,
+} from '@kubev2v/types';
+
+export type PowerState = 'on' | 'off' | 'unknown';
+
+// moved from packages/legacy/src/common/components/VMNameWithPowerState.tsx
+export const getVmPowerState = (
```

Comment:
```suggestion
export const getVmPowerState = (vm?: ProviderVirtualMachine): PowerState => {
  if (!vm) return 'unknown';
  
  const providerType = getProviderTypeFromVm(vm);
  
  switch (providerType) {
    case 'ovirt':
      return getOvirtVmPowerState(vm);
    case 'vsphere':
      return getVSphereVmPowerState(vm);
    case 'openstack':
      return getOpenstackVmPowerState(vm);
    case 'openshift':
      return getOpenshiftVmPowerState(vm);
    case 'ova':
      return getOvaVmPowerState();
    default:
      return 'unknown';
  }
};

const getOvirtVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const status = (vm as OVirtVM).status;
  if (status === 'up') return 'on';
  if (status === 'down') return 'off';
  return 'unknown';
};

const getVSphereVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const powerState = (vm as VSphereVM).powerState;
  if (powerState === 'poweredOn') return 'on';
  if (powerState === 'poweredOff') return 'off';
  return 'unknown';
};

const getOpenstackVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const status = (vm as OpenstackVM).status;
  if (status === 'ACTIVE') return 'on';
  if (status === 'SHUTOFF') return 'off';
  return 'unknown';
};

const getOpenshiftVmPowerState = (vm: ProviderVirtualMachine): PowerState => {
  const status = (vm as OpenshiftVM)?.object?.status?.printableStatus;
  return status === 'Running' ? 'on' : 'off';
};

const getOvaVmPowerState = (): PowerState => {
  return 'off';
};

const getProviderTypeFromVm = (vm: ProviderVirtualMachine): ProviderType => {
  // Assuming this method exists and returns the correct type
  return getProviderType(vm);
};
```

My concerns about the  way this method is currently implemented:
a. if you  need the type, then it's not a universal method, we may not need it if we know the type when we call it.
b. some anonymous methods extract the status into a const befor using it, and some not
c. each anonymous method actually has a name, each get the power for a different provider, they can be extracted to individual methods with names that explain what they do.

Reply:
p.s.

`getProviderTypeFromVm`  is not a real method, it's there to show how a universal method work, it should not need to get the type, if you need the type, you may not need this method at all, and you can just use the individual methods.

Reply:
p.p.s

you can add a providerType to the vm interface, or to the https://github.com/kubev2v/forklift-console-plugin/blob/main/packages/types/src/types/provider/openstack/Resource.ts interface (if it't the parent of all provider types) and use it to know the provider type, see:
https://github.com/kubev2v/forklift-console-plugin/blob/main/packages/types/src/types/secret/OvaProviderSecret.ts#L9C3-L9C23

Reply:
p.p.p.s
hmm, each provider will have different state nuances, best to return not just the state but alsoe the `icon`, `shortText` and `tooltip message`, so each provider can have it's own definition of what is on/off/migrating/sleeping etc. the text for openshift "on" maybe different from openstack "on" and the "off" for OVA will be different from vsphere "off"

Reply:
@yaacov 
>  if you need the type, then it's not a universal method, we may not need it if we know the type when we call it.
>

We need a way to chose between execution paths -  however if we extend the VM type to include `providerType` then we can branch on that.

> each provider will have different state nuances, best to return not just the state but alsoe the icon, shortText and tooltip message, so each provider can have it's own definition of what is on/off/migrating/sleeping etc. the text for openshift "on" maybe different from openstack "on" and the "off" for OVA will be different from vsphere "off"
>

We rely on common set of states (`PowerState` interface) for filtering. If we want to have different states then we need a power state enum per provider. If the differences are limited to the presentation layer (icon, tooltip,etc) then we can handle it in the renderer i.e.use different `states` dictionary. Currently a common mapping is used:
```
 const states: { [key in PowerState]: [JSX.Element, string, string] } = {
    on: [<PowerOffIcon color="green" key="on" />, t('Powered on'), t('On')],
    off: [<OffIcon color="red" key="off" />, t('Powered off'), t('Off')],
    unknown: [<UnknownIcon key="unknown" />, t('Unknown power state'), t('Unknown')],
  };
```

Reply:
Addressed in the newest version.